### DISCUSSION TITLE
time axis challenges: xcdat -> numpy

### DISCUSSION THREAD

**Discussion by gleckler1** [xCDAT/xcdat#438] [2. Questions] [answered]
[I am testing the use of xcdat to read in a large dataset via xcdat.open_mfdataset](https://github.com/PCMDI/obs4MIPs-cmor-tables/blob/245_pjg_hifqpr/inputs/pcmdi/GloH2O/MSWEP-V280/runCmor_MSWEP-v280.py).  My goal is to re-write this data using CMOR, but to do that I need to isolate the time axis as a numpy array with index values (with cdms this is rel time instead of component time).  I am struggling to find a reasonable/simple path.   Any hints?

**Comment by pochedls** [comment, marked as answer]
Does this work: `time = ds.time.values`?

**Reply by gleckler1** [reply]
>>> type(time)
<class 'xarray.core.dataarray.DataArray'>
>>> t4 = time.time.values
>>> type(t4)
<class 'numpy.ndarray'>
>>> t4[0:4]
array([cftime.DatetimeGregorian(1982, 1, 1, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1982, 1, 2, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1982, 1, 3, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1982, 1, 4, 0, 0, 0, 0, has_year_zero=False)],
      dtype=object)

SP, Thanks for the rapid response!  I am trying to translate the cftime.DatetimeGregorian values to an integer (days since datum).

**Reply by pochedls** [reply]
@gleckler1 – I think you have two options: 

1. Open the dataset with `ds = xc.open_mfdataset(fn, decode_times=False)` or 

2. try to re-encode the time axis after the fact:

```
from xarray.coding.times import encode_cf_datetime

units = ds.time.encoding['units']
calendar = ds.time.encoding['calendar']
time_decoded = encode_cf_datetime(ds.time, units, calendar)
print(time_decoded)
```

> (array([711399.5, 711429. , 711458.5, 711489. , 711519.5, 711550. , ...
>        734940.5, 734971.5, 735002. , 735032.5, 735063. , 735093.5]), 'days since 0001-01-01', 'noleap')

**Reply by gleckler1** [reply]
@pochedls That works great!  Now I need to do the same with the time_bnds - not obvious how to extend the solution.  I'll try striping out the lower bounds into one array and the upper into another, apply encode to both, and then reconstruct as tuple numpy.array.  Many thanks!

**Reply by tomvothecoder** [reply]
Thanks for providing a solution Steve. I'll mark this as the answer.

> @pochedls That works great! Now I need to do the same with the time_bnds - not obvious how to extend the solution. I'll try striping out the lower bounds into one array and the upper into another, apply encode to both, and then reconstruct as tuple numpy.array. Many thanks!

@gleckler1 Solution 1 will leave time and time bounds encoded, but you won't be able to perform temporal operations in xCDAT unless they are decoded. In Solution 2, `encode_cf_datetime()` seems to only operate on a 1-d array so your workaround sounds like the way to go.

_One thing to note is that `encode_cf_datetime()` probably won't work with non-CF compliant time units ("months since ..." or "years since ...")._

> Given an array of datetime objects, returns the tuple `(num, units, calendar)` suitable for a _CF compliant time variable_.
> Source - [xarray.coding.times.encode_cf_datetime docstring](https://github.com/pydata/xarray/blob/1e361ccb9123fe25acfd9e3364c911c1eec7d9db/xarray/coding/times.py#L613-L623)
